microbench/src/main/java/io/netty/microbench/concurrent/DefaultPromiseBenchmark.java

```diff
@@ -0,0 +1,141 @@
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+import org.openjdk.jmh.annotations.AuxCounters;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+@State(Scope.Benchmark)
+@Fork(1)
+@Threads(1)
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+public class DefaultPromiseBenchmark extends AbstractMicrobenchmark {
+    private ExecutorService executorService;
+    private EventExecutor executor;
+    private TestGenericFutureListener[] listeners;
+    private TestGenericFutureListener[] lastListeners;
+    private static AtomicLong doneCount = new AtomicLong();
+
+    @Param({ "1000000" })
+    public int numPromises;
+
+    @Param({ "1", "4" })
+    public int numListeners;
+
+    @Param({ "0", "4" })
+    public int lateListeners;
+
+    @AuxCounters
+    @State(Scope.Thread)
+    public static class DoneCounter {
+        @Setup(Level.Iteration)
+        public void clean() {
+            doneCount.set(0);
+        }
+
+        public long doneCount() {
+            return doneCount.get();
+        }
+    }
+
+    @Setup(Level.Trial)
+    public void setup() {
+        doneCount.set(0);
+        executorService = Executors.newSingleThreadExecutor();
+        executor = new DefaultEventExecutor(executorService);
+        listeners = new TestGenericFutureListener[numListeners];
+        for (int i = 0; i < numListeners; ++i) {
+            listeners[i] = new TestGenericFutureListener();
+        }
+        lastListeners = new TestGenericFutureListener[numListeners];
+        for (int i = 0; i < numListeners; ++i) {
+            lastListeners[i] = new TestGenericFutureListener();
+        }
+    }
+
+    @TearDown(Level.Trial)
+    public void teardown() {
+        try {
+            executorService.shutdown();
+        } finally {
+            executor.shutdownGracefully(0, 0, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @Benchmark
+    public void testInEventLoop(DoneCounter doneCounter) throws Exception {
+        CountDownLatch latch = new CountDownLatch(1);
+        startInEventLoop(latch);
+        latch.await();
+        Thread.sleep(1001);
+    }
+
+    @Benchmark
+    public void testOutEventLoop(DoneCounter doneCounter) throws Exception {
+        startOutEventLoop();
+        Thread.sleep(1001);
```

Comment:
> Do we really care if it's #/sec? Wouldn't a total/iteration be fine?

No. The intention is that we have some asynchronous operation running as the benchmark is executed. We want to give it some time to complete before executing the next execution of the benchmark. The rational for not using a `CountDownLatch` was that the CountDown latch itself could add non-trivial overhead to the test ... however if the test is overworked or under worked because the timeout is not perfect then this will also influence the test ... let me try a latch again.